docs: Add Component Overview topic to angular.io (#39186)
angular/angular: aio/content/guide/component-overview.md

You can define a template for your component in one of two ways: by referencing an external file, or directly within the component.

To define a template as an external file, add a `templateUrl` property to the `@Component` decorator.

<code-example
    path="component-overview/src/app/component-overview/component-overview.component.ts"
    region="templateUrl">
</code-example>

To define a template within the component, add a `template` property to the `@Component` decorator that contains the HTML you want to use.

<code-example
    path="component-overview/src/app/component-overview/component-overview.component.1.ts"
    region="template">
</code-example>

If your want your template to span multiple lines, you can use backticks (<code> ` </code>).
For example:

<code-example
    path="component-overview/src/app/component-overview/component-overview.component.2.ts"
    region="templatebacktick">
</code-example>

<div class="alert is-helpful">

An Angular component requires a template defined using `template` or `templateUrl`. You cannot have both statements in a component.

</div>

## Declaring a component's styles

> You can declar the styles your component uses for its template in one of two ways: by referencing an external file, or directly within the component.

Reviewer: ```suggestion
You can declare component styles in one of two ways: by referencing an external file, or directly within the component.
```
Potentially could make this slightly shorter since, if we assume baseline HTML and CSS knowledge, I would think that the relationship of the styles to the template is included in prior knowegde

> # Angular Components Overview

Reviewer: Is this doc meant to be an explanatory overview of "what is a component", or a tutorial? The name implies the former, but it seems structured as a tutorial.

Reviewer 2: It is a task-based guide. Task-based guides focus on how to get something done, but are not intended to be end-to-end solutions, which would make it a tutorial.

> An Angular application consists of up to four key parts: components, templates, directives, and dependencies.

Reviewer: I don't know that I would make this statement. The "up to", especially, doesn't exactly line up with the way that I would conceptually model an Angular app. A component _has_ a template, so treating them as separate constructs is kind of double counting. Components also _are_ directives (with some extra stuff). And "dependencies" here is a pretty broad term that could describe a lot of different things. 

Instead, I might say roughly something along the lines 
```markdown
Components are the main building block for Angular applications. 
Each component consists of:
* An HTML template that declares what renders on the page
* A TypeScript class that drives behavior
* A CSS selector that defines how the component is used in a template
* Optionally, CSS styles applied to the template
```

From there, I was thinking it might make the most sense to step through the anatomy of a component before getting into the Angular CLI commands to generate one. IMO it feels a bit early to dive into the scaffolding before getting into "what even is a component".

Reviewer 2: I don't mind updating the introduction. However, I don't want this topic to get into "What is a component?" per se. That content is conceptual, and should be covered in topics such as [Introduction to Angular Concepts](https://angular.io/guide/architecture).

Content in the guide should focus on how to get things done. That's why this topic focuses on how to create a component.

This topic describes Angular components: how to create them and how to configure them for your applications.

<div class="alert is-helpful">

To view or download the example code used in this topic, see the <live-example></live-example>.

</div>

## Creating a component

The easiest way to create a component is with the Angular CLI. You can also create a component manually.

> ### Creating a component using the Angular CLI

Reviewer: ```suggestion
### Scaffolding a component using the Angular CLI
```
Personally I would use the term "scaffolding here", though I could see how you might want to stick with a simpler term

Reviewer 2: I find Creating to be less jargon-y, to be honest.

To create a component using the Angular CLI, verify that you have met the following prerequisites:

1. Install the Angular CLI.
1. Create an Angular project.
   If you don't have a project, you can create one using `ng new <project-name>`, where `<project-name>` is the name of your Angular application.

> After you've met these prerequisites, you can create a new component by following these steps:

Reviewer: Are there any existing docs on getting started and/or creating projects with Angular CLI? Maybe a TODO for later, but it would be nice to have a more brief note here like `Note: this assume you have [a project using the Angular CLI](...)`

Reviewer 2: This information is covered here: [Setting up the local environment and workspace](https://angular.io/guide/setup-local), although we may want t separate topic that talks about creating a project.

By default, this command creates the following:

* A folder named after the component
* A component file, `<component-name>.component.ts`
* A template file, `<component-name>.component.ts`
* A CSS file, `<component-name>.component.css`
* A testing specification file, `<component-name>.component.spec.ts`

Where `<component-name>` is the name of your component.

<div class="alert is-helpful">

You can change how `ng generate component` creates new components.
For more information, see [ng generate component](cli/generate#component-command) in the Angular CLI documentation.

</div>

### Creating a component manually

Although the Angular CLI is the easiest way to create an Angular component, you can also create a component manually.
This section describes how to create the core component file within an existing Angular project.

As with creating a component using the Angular CLI, before you create a component you should verify that you have met the following prerequisites:

1. Install the Angular CLI.
1. Create an Angular project.
   If you don't have a project, you can create one using `ng new <project-name>`, where `<project-name>` is the name of your Angular application.

> After you've met these prerequisites, you can create a new component by following these steps:

Reviewer: Rather than repeating this, it seems like a prereq for the whole tutorial would just be having a CLI project already running.

# Angular Components Overview

> Components are the main building bloc for Angular applications. Each component consists of:

Reviewer: ```suggestion
Components are the main building block for Angular applications. Each component consists of:
```

        region="selector">
   </code-example>

   For more information on choosing a selector, see [Choosing a component's selector](#choosing-a-components-selector).

1. Define the HTML template that the component uses to display information.
   In most cases, this template is a separate HTML file.

   <code-example
        path="component-overview/src/app/component-overview/component-overview.component.ts"
        region="templateUrl">
   </code-example>

   For more information on defining a component's template, see [Defining a component's template](#defining-a-components-template).

1. Select the styles for the component's template.
   In most cases, you define the styles for you component's template in a separate file.

   <code-example
        path="component-overview/src/app/component-overview/component-overview.component.ts"
        region="decorator">
   </code-example>

1. Add a `class` statement that includes the code for the component.

   <code-example
        path="component-overview/src/app/component-overview/component-overview.component.ts"
        region="class">
   </code-example>

> ## Choosing a component's selector

Reviewer: The first time the term "selector" is used in this section, I would probably say "CSS selector" as the full term.

> Every component requires a _selector_. A selector instructs Angular to create and insert an instance of this component wherever it finds the corresponding tag in template HTML. For example, consider a component, `hello-world.component.ts` that defines its selector as `app-hello-world`. This selector instructs angular to instantiate this component any time the tag, `<app-hellow-world>` in a template.

Reviewer: ```suggestion
Every component requires a _selector_. A selector instructs Angular to instantiate this component wherever it finds the corresponding tag in template HTML. For example, consider a component, `login-form.component.ts` that defines its selector as `app-login-form`. This selector instructs Angular to instantiate this component any time the tag, `<app-login-form>` in a template.
```
few minor suggestions:
* I might condense "create and insert an instance of" down to just "instantiate" (since it's already using the term "instance" anyway)
* I might change the "hello-world" example here to be something more representative of real life.

Reviewer 2: I agree with instantiate!

I think hello-world is fine, but we can re-evaluate how we handle examples separately.

> To choose a component's selector, add a `selector` statement to the `@Component` decorator.

Reviewer: ```suggestion
To specify  a component's selector, add a `selector` statement to the `@Component` decorator.
```
I might use "specify" instead of "choose" here since (to me at least) "choose" implies a fixed set of options.

Reviewer 2: SGTM. Updated the heading to match.

<code-example
    path="component-overview/src/app/component-overview/component-overview.component.ts"
    region="selector">
</code-example>

> The recommended naming convention for a selector is to use `app-` followed by the name of the component. For example:

Reviewer: This actually isn't a recommended practice (as far as I know, anyway). "app" is just the default. I generally think people should choose a prefix based on their app/library. If I were in charge of the style guide, I'd actually give more nuanced guidance on selector conventions.

Reviewer 2: Really? I did not know that! I've always seen selector's named using that convention. To avoid confusion, I'm removing this sentence and the following code snippet.

```
app-component-overview
```

## Defining a component's template

A template is a block of HTML that tells Angular how to render the component in your application.
You can define a template for your component in one of two ways: by referencing an external file, or directly within the component.

To define a template as an external file, add a `templateUrl` property to the `@Component` decorator.

<code-example
    path="component-overview/src/app/component-overview/component-overview.component.ts"
    region="templateUrl">
</code-example>

To define a template within the component, add a `template` property to the `@Component` decorator that contains the HTML you want to use.

<code-example
    path="component-overview/src/app/component-overview/component-overview.component.1.ts"
    region="template">
</code-example>

If your want your template to span multiple lines, you can use backticks (<code> ` </code>).
For example:

<code-example
    path="component-overview/src/app/component-overview/component-overview.component.2.ts"
    region="templatebacktick">
</code-example>

<div class="alert is-helpful">

> An Angular component requires a template defined using `template` or `templateURL`. You cannot have both statements in a component.

Reviewer: ```suggestion
An Angular component requires a template defined using `template` or `templateUrl`. You cannot have both statements in a component.
```

</div>

> ## Selecting a component's styles

Reviewer: ```suggestion
## Declaring component styles
```
Again "selecting" isn't exactly the verb I would use for this (here and below)

You can select the styles your component uses for its template in one of two ways: by referencing an external file, or directly within the component.

> To define the styles for you component's template in a separate file, add a `stylesUrl` property to the `@Component` decorator.

Reviewer: ```suggestion
To define the styles for a component in a separate file, add a `styleUrls` property to the `@Component` decorator.
```

<code-example
    path="component-overview/src/app/component-overview/component-overview.component.ts"
    region="decorator">
</code-example>

To select the styles within the component, add a `styles` property to the `@Component` decorator that contains the styles you want to use.

<code-example
    path="component-overview/src/app/component-overview/component-overview.component.3.ts"
    region="styles">
</code-example>

> The `styles` property takes an array of strings that contain the CSS code.

Reviewer: ```suggestion
The `styles` property takes an array of strings that contain CSS rule declarations.
```
I generally wouldn't refer to CSS as "code", although one might debate this.
